### PR TITLE
New RTE full screen implementation with BottomSheet

### LINKED ISSUE #7577: Improve full screen mode for Rich Text Editor
#### What would you like to do?

Improve the UI and UX of the full screen mode for the Rich Text Editor.

#### Why would you like to do it?

The current implementation was provisional, as we had a tight deadline on delivering it. However, it does not match the designs and added unneeded complexity to `TimelineFragment` and `RichTextComposerLayout`, now having several constraint sets for the same layout.

#### How would you like to achieve it?

Use a custom `BottomSheetBehavior` implementation that resizes its contents and implementing missing designs.

### Have you considered any alternatives?

1. Creating a custom component not depending on `CoordinatorLayout`. This might make the implementation simpler at the cost at making it more difficult to re-use.
2. Keep using the current approach with constraint sets and applying new designs. However, the extra complexity of maintaining several layouts is still there.

### Additional context

_No response_

### Are you willing to provide a PR?

Yes

### PR DESCRIPTION
**Pull request by jmartinesp**
## Type of change

- [x] Feature
- [ ] Bugfix
- [ ] Technical
- [ ] Other :

## Content

* The RTE now uses a custom BottomSheet implementation instead of several constraint sets. This allows further customisations and improved UX.
* The full screen mode now covers the app bar and the shadow is applied to the status bar too, removing the previous weird looking white/black background.
* Shrank RTE min height when minimised.
* New icons for the attachment menu and voice recorder.
* New reply/edit/quote design for RTE, matching iOS.
* 'Send' button becomes 'Save' when editing a message.
* Added a 'minimise' counterpart to the 'maximise' icon.
* Simplified plain text layout quite a bit, removed the extra constraint sets so we rely on visibility changes instead.

## Motivation and context

Closes #7577 .

## Screenshots / GIFs

|Before|After|
|-|-|
|old_rte_fullscreen webm|new_rte_fullscreen webm|

## Tests

- Enable Rich Text Editor in Labs.
- Test full screen editing in several situations:
    - No text.
    - Edit / Reply / Quote enabled.
    - Voice recording.
    - Writing > 10 lines (maxLines).
    - Switching to plain text mode and back using the + button and toggling 'Rich text formatting'.
-  Disable RTE in Labs and test:
    - The layout resizes as needed, there is no full screen mode.
    - The old Edit/Reply/Quote layout is still present and working.
    - The editor can't be dragged up for full screen mode.

## Tested devices

- [ ] Physical
- [x] Emulator
- OS version(s): 9

## Checklist

- [ ] Changes has been tested on an Android device or Android emulator with API 21
- [x] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [x] Pull request is based on the develop branch
- [x] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [x] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [x] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

### CODE REVIEW

**Comment by Florian14** [comment]
Small bug:
When I open the actions menu with the keyboard expanded, the composer is still constrained by the keyboard position after closing the menu.

https://user-images.githubusercontent.com/11990514/201894850-f063c83c-f5d6-4a03-9b9f-f96f3ce86705.mp4

**Comment by Florian14** [comment]
Change the code style (bold/italic...) in the middle of a word is not possible anymore? It was working fine in my previous tests

**Comment by Florian14** [comment]
I expected to have the attachment icons below the editor when the text formatting is disabled. The empty area is a bit strange in my opinion. Not an issue, just personal feedback

**Comment by jmartinesp** [comment]
> Change the code style (bold/italic...) in the middle of a word is not possible anymore? It was working fine in my previous tests

There is an issue for that, the way Android IME report text changes sometimes breaks formatting for the word currently being written.

**Comment by jmartinesp** [comment]
> I expected to have the attachment icons below the editor when the text formatting is disabled. The empty area is a bit strange in my opinion. Not an issue, just personal feedback

That was done to meet the deadline, but there is already a ticket to re-arrange the composer layout when the plain text mode is enabled.

**Review by Florian14 — COMMENTED** [review]
2 more bugs:

1. On real device, the editor is not correctly reset after sending a message (I don't have the problem on the emulator). After the first message the send button disappear, after adding a new character, I can send the message again and the button does not disappear anymore.
2. After a screen rotation, I lost the line breaks on the text formatting

Another point is that the editor is shown behind the status bar, is it normal? The rendering is not fine from my point of view. And I was not able to scroll in the composer in landscape mode.

https://user-images.githubusercontent.com/11990514/201946675-b209f45c-a458-4bd7-8535-bbcea8d69156.mov

**Comment by Florian14** [comment]
Not sure the bugs are introduced by this PR, the code looks fine but it was hard to understand all the logic.
Great work otherwise! It's promising 👍🏻

**Comment by jmartinesp** [comment]
@Florian14 the 1st and 3rd issues you mentioned should be fixed in c1af6d9 and bd8f83a. I confirmed that the 2nd one is coming from the RTE library, and should be fixed with a new release.

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -7,7 +7,7 @@ on:
 
 # Enrich gradle.properties for CI/CD
 env:
-  GRADLE_OPTS: -Dorg.gradle.jvmargs="-Xmx3072m -Dfile.encoding=UTF-8 -XX:+HeapDumpOnOutOfMemoryError" -Dkotlin.daemon.jvm.options="-Xmx2560m" -Dkotlin.incremental=false
+  GRADLE_OPTS: -Dorg.gradle.jvmargs="-Xmx3584m -Dfile.encoding=UTF-8 -XX:+HeapDumpOnOutOfMemoryError" -Dkotlin.daemon.jvm.options="-Xmx2560m" -Dkotlin.incremental=false
```
[the remark this answers is not in the document]
Comment (jmartinesp) [reply .github/workflows/tests.yml:10]: The UI tests were failing before with OOM, so I increased it to try to make them work again.

**Pull request merged**